src/etherscan/EtherscanVerifyContractRequest.ts

```diff
@@ -0,0 +1,67 @@
+import { BuidlerPluginError } from "@nomiclabs/buidler/plugins";
+import { SolcConfig } from "@nomiclabs/buidler/types";
+
+import { EtherscanBuidlerEnvironment } from "../index";
+
+export default class EtherscanVerifyContractRequest {
+  // for that weird etherscan library props
+  [key: string]: any;
+  public readonly apikey: string;
+  public readonly module: string = "contract";
+  public readonly action: string = "verifysourcecode";
+  public readonly contractaddress: string;
+  public readonly sourceCode: string;
+  public readonly contractname: string;
+  public readonly compilerversion: string;
+  public readonly optimizationUsed: number;
+  public readonly runs: number;
+  public readonly constructorArguements: string;
+
+  constructor(
+    etherscanConfig: EtherscanBuidlerEnvironment,
+    solcConfig: SolcConfig,
+    contractName: string,
+    address: string,
+    libraries: string,
+    source: string,
+    constructorArguments: string
+  ) {
+    this.apikey = etherscanConfig.token;
+    this.contractaddress = address;
+    this.sourceCode = source;
+    this.contractname = contractName;
+    this.compilerversion = solcConfig.fullVersion;
+    this.optimizationUsed = solcConfig.optimizer.enabled ? 1 : 0;
+    this.runs = solcConfig.optimizer.runs;
+    this.constructorArguements = constructorArguments;
+    this.setLibraries(libraries);
+  }
+
+  public serialize(): string {
+    return JSON.stringify(this);
+  }
+
+  private setLibraries(libraries: string) {
+    let i: number = 1;
+    let parsedLibraries: { [key: string]: string } = {};
+    try {
+      if (libraries) {
+        parsedLibraries = JSON.parse(libraries);
+      }
+    } catch (e) {
+      throw new BuidlerPluginError(
+        "Failed to parse libraries. Reason: " + e.message
+      );
+    }
+    for (const libraryName in parsedLibraries) {
+      if (parsedLibraries.hasOwnProperty(libraryName)) {
+        this["libraryname" + i] = libraryName;
+        this["libraryaddress" + i] = parsedLibraries[libraryName];
+        i++;
+        if (i >= 10) {
```

Comment:
Why is this needed?

Reply:
Yeah, weird as hell. I don't think it's working via API, since it fails on their client as well but works when submitted trough GUI. I'm trying to find where I can report this.